File: UI_MOdified/server/tile-server.js
/**
 * MBTiles tile server - uses better-sqlite3 to read from disk (supports large files, e.g. 25GB+).
 * Run from project root: node server/tile-server.js  OR  npm run start
 */
const express = require('express');
const path    = require('path');
const fs      = require('fs');

let Database;
try {
    Database = require('better-sqlite3');
} catch (e) {
    console.error('\n  ERROR: better-sqlite3 not installed. Run: npm install\n');
    process.exit(1);
}

const app      = express();
/** maps/ folder — env var (set by Electron) or project-root default */
const MAPS_DIR = process.env.RMOOZ_MAPS_DIR || path.join(__dirname, '..', 'maps');
const PORT     = parseInt(process.env.TILE_SERVER_PORT, 10) || 8080;

// Allow the web page (port 8000) to call this server
app.use((req, res, next) => {
    res.set('Access-Control-Allow-Origin', '*');
    next();
});

const dbs = {};  // tileset name → { db, format }

function loadAll() {
    if (!fs.existsSync(MAPS_DIR)) {
        console.warn('  WARNING: maps/ folder not found at', MAPS_DIR);
        return;
    }

    const files = fs.readdirSync(MAPS_DIR).filter(f => f.toLowerCase().endsWith('.mbtiles'));

    if (files.length === 0) {
        console.warn('\n  WARNING: No .mbtiles files found in maps/');
        console.warn('  Copy your .mbtiles file into the maps/ folder.\n');
        return;
    }

    for (const filename of files) {
        const filepath = path.join(MAPS_DIR, filename);
        const name     = filename.replace(/\.mbtiles$/i, '');
        const sizeMB   = (fs.statSync(filepath).size / (1024 * 1024)).toFixed(1);
        console.log('  Loading:', filename, '(' + sizeMB + ' MB)...');
        try {
            const db = new Database(filepath, { readonly: true });
            // Read the full metadata table so we can expose real minzoom/maxzoom
            // (used by the client to set maxNativeZoom → overzoom instead of blank
            // tiles when zooming past the data ceiling).
            const meta = {};
            try { for (const r of db.prepare('SELECT name, value FROM metadata').all()) meta[r.name] = r.value; } catch (_) {}
            const fmt = (meta.format) || 'png';
            let minzoom = parseInt(meta.minzoom, 10);
            let maxzoom = parseInt(meta.maxzoom, 10);
            if (!Number.isFinite(maxzoom)) { try { maxzoom = db.prepare('SELECT MAX(zoom_level) z FROM tiles').get().z; } catch (_) {} }
            if (!Number.isFinite(minzoom)) { try { minzoom = db.prepare('SELECT MIN(zoom_level) z FROM tiles').get().z; } catch (_) {} }
            dbs[name] = { db, format: fmt, minzoom, maxzoom, bounds: meta.bounds || null, attribution: meta.attribution || null };
            console.log('  Loaded :', name, '(format:', fmt, '| zoom', minzoom + '..' + maxzoom + ')');
        } catch (e) {
            console.error('  FAILED :', filename, '-', e.message);
        }
    }
}

// TileJSON-ish metadata: lets the client read the tileset's real maxzoom and set
// Leaflet maxNativeZoom, so zooming past the data ceiling upscales (overzoom)
// instead of requesting non-existent tiles and showing blank/grey.
app.get('/services/:tileset.json', (req, res) => {
    const entry = dbs[req.params.tileset];
    if (!entry) { res.status(404).json({ error: 'unknown tileset' }); return; }
    res.json({
        tilejson: '2.2.0',
        name:     req.params.tileset,
        format:   entry.format,
        scheme:   'xyz',
        minzoom:  Number.isFinite(entry.minzoom) ? entry.minzoom : 0,
        maxzoom:  Number.isFinite(entry.maxzoom) ? entry.maxzoom : 22,
        bounds:   entry.bounds ? entry.bounds.split(',').map(Number) : undefined,
        attribution: entry.attribution || undefined,
    });
});

app.get('/services/:tileset/:z/:x/:y.:fmt', (req, res) => {
    const { tileset, z, x, y } = req.params;
    const entry = dbs[tileset];
    if (!entry) { res.status(404).send(); return; }

    try {
        const maxY = Math.pow(2, parseInt(z, 10)) - 1;
        const tmsY = maxY - parseInt(y, 10);
        const row = entry.db.prepare(
            'SELECT tile_data FROM tiles WHERE zoom_level=? AND tile_column=? AND tile_row=?'
        ).get(parseInt(z), parseInt(x), tmsY);

        if (row && row.tile_data) {
            const fmt = (entry.format || 'png').toLowerCase();
            res.set('Content-Type', fmt === 'jpg' || fmt === 'jpeg' ? 'image/jpeg' : 'image/png');
            res.send(Buffer.from(row.tile_data));
        } else {
            res.status(404).send();
        }
    } catch (e) {
        res.status(500).send();
    }
});

loadAll();
const tileListener = app.listen(PORT, '127.0.0.1', () => {
    console.log('');
    console.log('  Tile server running at http://127.0.0.1:' + PORT);
    console.log('  Tilesets ready:', Object.keys(dbs).join(', ') || '(none - check maps/ folder)');
    console.log('  (Supports large files - reads from disk, no size limit)');
    console.log('');
});
tileListener.on('error', (err) => {
    if (err && err.code === 'EADDRINUSE') {
        console.error('Port ' + PORT + ' is already in use. Stop the other tile server (e.g. lsof -ti tcp:' + PORT + ' | xargs kill) or change PORT in server/tile-server.js.');
    } else {
        console.error(err);
    }
    process.exit(1);
});

File: UI_MOdified/server/tile-server.py
#!/usr/bin/env python3
"""Simple MBTiles tile server - no npm/Node required. Uses Python's built-in sqlite3.
Run from project root: python server/tile-server.py
"""
import http.server
import sqlite3
import urllib.parse
import os

PORT = 8080
# maps/ folder lives at the project root (parent of server/)
MAPS_DIR = os.path.join(os.path.dirname(os.path.abspath(__file__)), '..', 'maps')

_format_cache = {}

def get_tile_format(tileset):
    if tileset in _format_cache:
        return _format_cache[tileset]
    tile_path = os.path.join(MAPS_DIR, tileset + '.mbtiles')
    fmt = 'png'
    try:
        conn = sqlite3.connect(tile_path)
        cur = conn.cursor()
        cur.execute("SELECT value FROM metadata WHERE name='format'")
        row = cur.fetchone()
        if row and row[0]:
            fmt = str(row[0]).lower()
        conn.close()
    except Exception:
        pass
    _format_cache[tileset] = fmt
    return fmt

def get_tile(tileset, z, x, y):
    tile_path = os.path.join(MAPS_DIR, tileset + '.mbtiles')
    if not os.path.exists(tile_path):
        return None
    try:
        conn = sqlite3.connect(tile_path)
        conn.row_factory = sqlite3.Row
        cur = conn.cursor()
        max_y = (1 << int(z)) - 1
        tms_y = max_y - int(y)
        cur.execute(
            'SELECT tile_data FROM tiles WHERE zoom_level=? AND tile_column=? AND tile_row=?',
            (int(z), int(x), tms_y)
        )
        row = cur.fetchone()
        conn.close()
        return row['tile_data'] if row else None
    except Exception:
        return None

class TileHandler(http.server.BaseHTTPRequestHandler):
    def do_GET(self):
        parts = self.path.strip('/').split('/')
        if len(parts) >= 5 and parts[0] == 'services':
            tileset, z, x, y_ext = parts[1], parts[2], parts[3], parts[4]
            y = y_ext.rsplit('.', 1)[0] if '.' in y_ext else y_ext
            data = get_tile(tileset, z, x, y)
            if data:
                fmt = get_tile_format(tileset)
                ctype = 'image/jpeg' if fmt in ('jpg', 'jpeg') else (
                    'image/webp' if fmt == 'webp' else 'image/png'
                )
                self.send_response(200)
                self.send_header('Content-Type', ctype)
                self.send_header('Access-Control-Allow-Origin', '*')
                self.end_headers()
                self.wfile.write(data)
                return
        self.send_response(404)
        self.end_headers()

    def log_message(self, format, *args):
        pass

def main():
    # Run with the project root as cwd so relative imports stay consistent
    project_root = os.path.join(os.path.dirname(os.path.abspath(__file__)), '..')
    os.chdir(project_root)
    mbtiles = [f[:-9] for f in os.listdir(MAPS_DIR) if f.endswith('.mbtiles')]
    print('\nMBTiles tile server at http://localhost:%d' % PORT)
    print('Tilesets:', ', '.join(mbtiles) or '(none)')
    print('\nRun start-server.bat in another window, then open http://localhost:8000\n')
    server = http.server.HTTPServer(('', PORT), TileHandler)
    server.serve_forever()

if __name__ == '__main__':
    main()
